### ISSUE TITLE
Stop Engine from blocking

### ISSUE THREAD

**Issue by jeswr** [comunica/comunica#995]
#### Issue type:

- :snail: Performance issue 

____
#### Description:

As identified in [this PR](https://github.com/comunica/comunica/pull/981#issuecomment-1110568121) there are some points at which the engine will block, such as when `await`ing metadata from an iterator. IMO any component that does this should not *block* and rather, return an `AsyncIterable` iterator that is not yet readable.

The consequence of this is that calling `engine.execute()` should *never* cause blocking behavior, but simply return an iterator that will become readable *once* data is ready.

____

**Comment by jeswr** [comment]
To add on to this mediators should not block necessarily. For instance, should not wait on all tests if one returns a cost of 0 `MediatorJoinCoefficientsFixed`.

**Comment by jeswr** [comment]
Note that the promise wrapper introduced in https://github.com/RubenVerborgh/AsyncIterator/pull/63 will probably help with this

**Comment by rubensworks** [comment]
Oh, good catch @jeswr.

> The consequence of this is that calling engine.execute() should never cause blocking behavior, but simply return an iterator that will become readable once data is ready.

👍 

> To add on to this mediators should not block necessarily. For instance, should not wait on all tests if one returns a cost of 0 MediatorJoinCoefficientsFixed.

Sounds like a good solution for this specific problem!

**Comment by jeswr** [comment]
> Sounds like a good solution for this specific problem!

Ive started on this fix in the promise-source-debugging branch. Will clean it up and make a separate PR when I get the chance

**Comment by jacoscaz** [comment]
> The consequence of this is that calling engine.execute() should never cause blocking behavior, but simply return an iterator that will become readable once data is ready.

I generally agree on this but it's worth pointing out that there's a tradeoff between blocking behavior and shorter chains of iterators.  https://github.com/RubenVerborgh/AsyncIterator/pull/63 makes wrapping promisified iterators a lot faster but it doesn't entirely eliminate wrapping overhead as we still end up with one additional "pass-through" iterator that becomes readable once the promise resolves to the source iterator. Whereas the overhead is minimal with synchronous-ish sources, it _might_ become more significant with async sources emitting numerous `readable` events due to https://github.com/RubenVerborgh/AsyncIterator/blob/fec14f4d19e3d594bc385b927863ebf80a792580/asynciterator.ts#L258 .

Nonetheless, I do agree that making `execute()` as non-blocking as possible should be a priority.

**Comment by rubensworks** [comment]
I just thought about this some more, and actually await-ing metadata should be ok, since we assume that metadata is calculated quickly (it's in the docs somewhere).
So if blocking metadata calls really stall execution, then there's a problem somewhere in the logic that produces that metadata.
